lib/cartopy/mpl/geoaxes.py

```diff
@@ -1649,213 +1659,171 @@ def pcolormesh(self, *args, **kwargs):
+        # Add in an argument checker to handle Matplotlib's potential
+        # interpolation when coordinate wraps are involved
+        args = self._wrap_args(*args, **kwargs)
+        result = matplotlib.axes.Axes.pcolormesh(self, *args, **kwargs)
+        # Wrap the quadrilaterals if necessary
+        result = self._wrap_quadmesh(result, **kwargs)
+        # Re-cast the QuadMesh as a GeoQuadMesh to enable future wrapping
+        # updates to the collection as well.
+        result.__class__ = cartopy.mpl.geocollection.GeoQuadMesh
+
         self.autoscale_view()
         return result
 
-    def _pcolormesh_patched(self, *args, **kwargs):
-        """
-        A modified duplicate of :func:`~matplotlib.pyplot.pcolormesh'.
-
-        This function contains patches for Cartopy-specific behaviour, such as
-        using the transform for limit checks, applying longitude wrapping, etc.
-        See PATCH comments below.
-
+    def _wrap_args(self, *args, **kwargs):
+        """
+        Handle the interpolation when a wrap could be involved with
+        the data coordinates before passing on to Matplotlib.
+        """
+        if (kwargs.get('shading', 'auto') in ('nearest', 'auto') and
+                len(args) == 3 and
+                isinstance(kwargs.get('transform'), _WRAP_PROJECTIONS)):
+            X = np.asanyarray(args[0])
+            Y = np.asanyarray(args[1])
+            nrows, ncols = np.asanyarray(args[2]).shape
+            Nx = X.shape[-1]
+            Ny = Y.shape[0]
+            if X.ndim != 2 or X.shape[0] == 1:
+                x = X.reshape(1, Nx)
+                X = x.repeat(Ny, axis=0)
+            if Y.ndim != 2 or Y.shape[1] == 1:
+                y = Y.reshape(Ny, 1)
+                Y = y.repeat(Nx, axis=1)
+
+            def _interp_grid(X, wrap=0):
+                # helper for below
+                if np.shape(X)[1] > 1:
+                    dX = np.diff(X, axis=1)
+                    # account for the wrap
+                    if wrap:
+                        dX = (dX + wrap/2) % wrap - wrap/2
+                    dX = dX/2
+                    X = np.hstack((X[:, [0]] - dX[:, [0]],
+                                   X[:, :-1] + dX,
+                                   X[:, [-1]] + dX[:, [-1]]))
+                else:
+                    # This is just degenerate, but we can't reliably guess
+                    # a dX if there is just one value.
+                    X = np.hstack((X, X))
+                return X
+            t = kwargs.get('transform')
+            xwrap = abs(t.x_limits[1] - t.x_limits[0])
+            if ncols == Nx:
+                X = _interp_grid(X, wrap=xwrap)
+                Y = _interp_grid(Y)
+            if nrows == Ny:
+                X = _interp_grid(X.T, wrap=xwrap).T
+                Y = _interp_grid(Y.T).T
+
+            args = (X, Y, args[2])
+        return args
+
+    def _wrap_quadmesh(self, collection, **kwargs):
```

Comment:
This looks nice - is it clear that it covers changing the central longitude while preserving as much data as possible?

Reply:
Good question. `central_longitude` or any center point shouldn't come into play here. This "wrapping" function works on the transformed coordinates, identifying when those are large, so, just how wide the `x_limits` are to identify the masked areas.

For the other new function to do the edge interpolation:
We are identifying which direction the subtraction should be with the modulo arithmetic. i.e. is 1 -> 359 a difference of 358 or 2. Before it was 358 and a large jump, now we're saying it is 2. So again, no worries about central longitude. The central longitude is handled in the CRS/PROJ routines.